### ISSUE TITLE
Docker node libcurl4 issues?

### ISSUE THREAD

**Issue by vck3000** [nodkz/mongodb-memory-server#313]
## Versions

- NodeJS: 13
- mongodb-memory-server-*: 6.5.2
- mongodb: latest
- system: Docker node:13

package: mongo-memory-server

## What is the Problem?
I have an integration test set that involves docker-compose creating two containers - a redis container and a backend container that uses mongo-memory-server.

In the past I have had no issues with this setup. However I recently reinstalled my OS (Ubuntu 20.04) and have now been having this weird issue since:

```
backend-testing    | 2020-06-05T08:32:29.251Z MongoMS:MongoMemoryServer Called MongoMemoryServer.ensureInstance() method:
backend-testing    | 2020-06-05T08:32:29.252Z MongoMS:MongoMemoryServer  - no running instance, call `start()` command
backend-testing    | 2020-06-05T08:32:29.252Z MongoMS:MongoMemoryServer Called MongoMemoryServer.start() method
backend-testing    | 2020-06-05T08:32:29.256Z MongoMS:MongoMemoryServer Starting MongoDB instance with following options: {"port":45885,"dbName":"f900b45c-d337-4539-aed9-95818b874c3a","ip":"127.0.0.1","storageEngine":"ephemeralForTest","dbPath":"/tmp/mongo-mem-30Vc6lQvzFiMEB","tmpDir":{"name":"/tmp/mongo-mem-30Vc6lQvzFiMEB"},"uri":"mongodb://127.0.0.1:45885/f900b45c-d337-4539-aed9-95818b874c3a?"}
backend-testing    | 2020-06-05T08:32:29.258Z MongoMS:MongoBinary MongoBinary options: {"downloadDir":"/usr/src/app/packages/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries","platform":"linux","arch":"x64","version":"4.0.14"}
backend-testing    | 2020-06-05T08:32:29.264Z MongoMS:MongoBinary MongoBinary: Mongod binary path: /usr/src/app/packages/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.14/mongod
backend-testing    | 2020-06-05T08:32:29.283Z MongoMS:MongoInstance Mongo[45885]: Called MongoInstance._launchKiller(parent: 30, child: 43):
backend-testing    | 2020-06-05T08:32:29.301Z MongoMS:MongoBinary MongoBinary: Download lock removed
backend-testing    | 2020-06-05T08:32:29.302Z MongoMS:MongoInstance Mongo[45885]: STDERR: /usr/src/app/packages/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.14/mongod: /usr/lib/x86_64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_4' not found (required by /usr/src/app/packages/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.14/mongod)
backend-testing    | 
backend-testing    | 2020-06-05T08:32:29.302Z MongoMS:MongoInstance Mongo[45885]: CLOSE: 1
backend-testing    | 2020-06-05T08:32:31.341Z MongoMS:MongoInstance Mongo[45885]: [MongoKiller]: exit - [0,null]
```

The weirdest part is that it still runs perfectly fine on my CI builds, but not locally in docker. 

## Code Example
Here is my dockerfile. The CI dockerfile is identical.

```
FROM node:13

# Setting working directory. All the path will be relative to WORKDIR
WORKDIR /usr/src/app

# Copy over required files
# For now copy everything. Can optimise this by not copying frontend since 
# there are no integration tests for it.
COPY ./ ./

# Install packages
RUN npx lerna bootstrap
RUN npm run build:proto

# Dev mode
CMD [ "npm", "run", "test:int:run" ]
```

Here is the docker-compose.yml file. The CI docker-compose file is identical except the two shared volumes are removed. 
```
version: '3'
services:
  redis-testing:
    image: redis:5.0.7-alpine # Use stable version
    container_name: redis-testing

  backend-testing:
    container_name: backend-testing
    depends_on:
      - redis-testing
    build:
      context: .
      dockerfile: Dockerfile.test
      args:
        SERVICE: backend
    environment:
      SERVICE: backend
      MONGOMS_DEBUG: 1
    volumes:
      - ./packages/backend/src:/usr/src/app/packages/backend/src
      - ./packages/proto/build:/usr/src/app/packages/proto/build
```

I'm using it like so in my Jest tests:
```
    // Set up database
    mongoServer = new MongoMemoryServer();
    const mongoUri = await mongoServer.getUri();
```

Full source code can be seen here: https://github.com/vck3000/ProAvalon/tree/newPA_staging 

## Do you know *why* it happens?

When I go into the docker image's files, the file `/usr/lib/x86_64-linux-gnu/libcurl.so.4` exists in the expected location. However I'm not sure whether it is the correct version. 

Creating a container of the image and exploring the packages installed, I found:
```
# dpkg -l | grep libcurl
ii  libcurl3:amd64                     7.52.1-5+deb9u10                  amd64        easy-to-use client-side URL transfer library (OpenSSL flavour)
ii  libcurl3-gnutls:amd64              7.52.1-5+deb9u10                  amd64        easy-to-use client-side URL transfer library (GnuTLS flavour)
ii  libcurl4-openssl-dev:amd64         7.52.1-5+deb9u10                  amd64        development files and documentation for libcurl (OpenSSL flavour)
```
It appears to be installed...

I'm pretty lost... :(. Any ideas?

**Comment by hasezoey** [comment]
the issue seems to be that you copy **everything**, including `node_modules` from you host, which also copies the downloaded version of `mongod`, which on you host requires libcurl4, but the docker container is different
an easy fix would be to add `RUN rm -rf node_modules && npm i` after your `COPY`, but the better way would be to just copy the needed files and then run `npm i`

**Comment by vck3000** [comment]
Wow, great catch. I had set up a `.dockerignore` file that should have prevented copying over all the `node_modules` but I'm not sure why that didn't work. I'll need to look into that. 

I removed the `node_modules/.cache` folder and reinstalled npm packages as you suggested and it now works perfectly again. Thanks for your help :).

Here is my `.dockerignore` if you were curious: 
```
*
!packages/backend/*
!packages/frontend/*
!packages/proto/*
packages/backend/node_modules/
packages/frontend/node_modules/
packages/proto/node_modules/
!./**/package-lock.json
!./**/package.json
!lerna.json
node_modules/
```
